### DISCUSSION TITLE
recent package upgrade causing a breakage

### DISCUSSION THREAD

**Discussion by tommyjcarpenter** [apache/airflow#21675] [General]
### Apache Airflow version

2.2.3 (latest released)

### What happened

`itsdangerous` had a release 16 hours ago, but I'm not 100% sure it's the cause of this. 
I see that https://pypi.org/project/apache-airflow-providers-cncf-kubernetes/ also had a release, 5 days ago, but I think this just started failing last night:

```


==================================== ERRORS ====================================
--
900 | _______ ERROR collecting tests/test_airflow_generate_indexapi_runner.py ________
901 | /root/.pyenv/versions/3.10.2/lib/python3.10/logging/config.py:390: in resolve
902 | found = getattr(found, frag)
903 | E   AttributeError: module 'airflow.utils.log' has no attribute 'file_task_handler'
904 |  
905 | During handling of the above exception, another exception occurred:
906 | /root/.pyenv/versions/3.10.2/lib/python3.10/logging/config.py:392: in resolve
907 | self.importer(used)
908 | .tox/unit/lib/python3.10/site-packages/airflow/utils/log/file_task_handler.py:25: in <module>
909 | from itsdangerous import TimedJSONWebSignatureSerializer
910 | E   ImportError: cannot import name 'TimedJSONWebSignatureSerializer' from 'itsdangerous' (/codebuild/output/src517196057/src/github.com/merqurian/python-indexapi/.tox/unit/lib/python3.10/site-packages/itsdangerous/__init__.py)
911 |  
912 | The above exception was the direct cause of the following exception:
913 | /root/.pyenv/versions/3.10.2/lib/python3.10/logging/config.py:565: in configure
914 | handler = self.configure_handler(handlers[name])
915 | /root/.pyenv/versions/3.10.2/lib/python3.10/logging/config.py:723: in configure_handler
916 | klass = self.resolve(cname)
917 | /root/.pyenv/versions/3.10.2/lib/python3.10/logging/config.py:399: in resolve
918 | raise v
919 | /root/.pyenv/versions/3.10.2/lib/python3.10/logging/config.py:392: in resolve
920 | self.importer(used)
921 | .tox/unit/lib/python3.10/site-packages/airflow/utils/log/file_task_handler.py:25: in <module>
922 | from itsdangerous import TimedJSONWebSignatureSerializer
923 | E   ValueError: Cannot resolve 'airflow.utils.log.file_task_handler.FileTaskHandler': cannot import name 'TimedJSONWebSignatureSerializer' from 'itsdangerous' (/codebuild/output/src517196057/src/github.com/merqurian/python-indexapi/.tox/unit/lib/python3.10/site-packages/itsdangerous/__init__.py)
```

here was the test trace that caused this:

```
The above exception was the direct cause of the following exception:
tests/test_airflow_generate_indexapi_runner.py:5: in <module>
    from airflow.models.dagbag import DagBag
.tox/unit/lib/python3.10/site-packages/airflow/__init__.py:46: in <module>
    settings.initialize()
.tox/unit/lib/python3.10/site-packages/airflow/settings.py:492: in initialize
    LOGGING_CLASS_PATH = configure_logging()
.tox/unit/lib/python3.10/site-packages/airflow/logging_config.py:74: in configure_logging
    raise e
.tox/unit/lib/python3.10/site-packages/airflow/logging_config.py:69: in configure_logging
    dictConfig(logging_config)
../../.pyenv/versions/3.10.2/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/logging/config.py:810: in dictConfig
    dictConfigClass(config).configure()
../../.pyenv/versions/3.10.2/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/logging/config.py:572: in configure
    raise ValueError('Unable to configure handler '
E   ValueError: Unable to configure handler 'task'
```

### What you expected to happen

this was working until yesterday, some recent release has broken airflow

### How to reproduce
We have a test that imports:

```
from airflow.models.dagbag import DagBag
```

our tox.ini just installs:

```
apache-airflow
apache-airflow-providers-cncf-kubernetes
```

This import now explodes

### Operating System

mac and ubuntu

### Versions of Apache Airflow Providers

```
apache-airflow
apache-airflow-providers-cncf-kubernetes
```

latest in pypi as of today

### Deployment

Other

### Deployment details

this was not in a deployment, just in a test, our test import the latest from pypi of :

```
apache-airflow
apache-airflow-providers-cncf-kubernetes
```

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
You should use constraints to install airflow : https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html 

Did you do it ? If not - can you make sure you do it (this is the only supported and recommended way of installing airflow that you should follow).

**Comment by potiuk** [comment]
Let us know if the only supported way of installing airflow works for you (I converted it into discussion).

**Comment by tommyjcarpenter** [comment]
I am installing this package via tox - and have to investigate if this is even possible. We are using this with tox/pytest and it just takes a list of package names currently. And we just import a single function - DagBAG.  So I have to research how to do this in our tests and ill report back.

**Reply by potiuk** [reply]
The only supported instalation is via PIP with constraints. if you are using your own methods, not following our recommendations - you are on your own, and it can get broken. You are responsible for the choice of your tools, not us.

**Reply by sprckt** [reply]
Hi @potiuk is support for installation via Poetry on your roadmap?

**Reply by potiuk** [reply]
Not from my side until they support constraints: https://github.com/python-poetry/poetry/issues/3225#issuecomment-749783200
The issue is closed even if some people claimed that they attempted to work on it. I have not heard anything since (and I am checking from time to time). I stated my opinion there which has 13 upvotes so far (feel free to make it 14) but apparently it is not a priority for them,

In any case - if someone will figure out how to make poetry installation with our constraint mechanism - i.e. wants to write a documentation and instructions (and support poetry users who will have problems) - feel free to do it. 

Airflow has > 2000 contributors, so if someone will provide good quality documentation handling all the edge cases and will supportt the users of poetry - there is nothing to stop anyone from doing it. I am happy to review/merge it providing that it will be from someone considering seriously all the cases.

**Reply by tommyjcarpenter** [reply]
FWIW that issue links to an Openstack example, and their example uses tox. 
I mentioned we use tox above - so clearly tox is fine since it is just "pip with constraints".

**Reply by sprckt** [reply]
Thanks for the quick reply. I upvoted your comment - let's see if it has any effect. We are currently trialing Airflow for our data pipelines and will try to dedicate some time to see if we can achieve install using Poetry.

**Reply by potiuk** [reply]
@sprckt Tox or poetry  it does not matter - as long as you do not use constraints, then yeah broken installation because someone else released a new version might happen.

Airflow has 600+ dependencies in total which means that every day 3-4 of them get updated on average. And yeah, new release any of those deps **might** break the installation at any time. We had this a number of times in the past  - including Werkzeug release breaking our installation 2 hours after we released important airflow version. 

This is PRECISELY why we introduced constraints. Constraints provide a mechanism where ALL 600 dependencies are specified in the exact version that they can be installed without conficts. At the same time constraints (unlike pinned install-required dependencies) do not block the users from upgrading dependencies if they find a neeed to do so. You can read more about it here https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pinned-constraint-files - it explains why all of that is needed.

If you want to install airflow in repeatable way - use [constraints](https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html) - whether it's `poetry`, or  `pip` or anythong else - I do not care. But currently only `pip` supports constraints natively (as far as I know, and I am following this up rather closely), so it is the only way we consider as 'good'. If you can work out and contribute a repeatable, sustainable way that our constraints can be used during poetry installation - by all means pleas do and contribute the documentation on how to do it back. I would love to be able to support poetry installations, simply the "straightforward" way (i.e. poetry supporting constraints) has not been followed by the poetry maintainers.

**Reply by tommyjcarpenter** [reply]
Having 600 dependencies in any project sounds like the source of much frustration.

**Reply by potiuk** [reply]
> Having 600 dependencies in any project sounds like the source of much frustration.

It's not. I's just the matter of using constraints.

**Reply by potiuk** [reply]
If you want, you can watch my talk about it I gave few months ago - why we need it and why it makes sense for us and how we solved it https://www.youtube.com/watch?v=_SjMdQLP30s&t=2549s (with constraints).

**Reply by potiuk** [reply]
And it's also the reason of our success - we are the most contributed to ASF project and the most successufl orchestrator, ubiquitous out there. And yes. We have AIPs https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals (including multi-tenancy and runtime isolation) that will make it modularisation more possible, even though Python itself does not allow separate venvs for different "integrations" yet - including separate Docker runtimes and such.

But until this happens - we have sophisticated mechanism to work out a "good" constraint set - and the best choice is to, well, use them maybe, especially that this is the only way we support :D.

**Reply by potiuk** [reply]
Any contributions to improve it are most welcome BTW.

**Comment by tommyjcarpenter** [comment]
@potiuk what does the constraint file dictate - in the sense that, is 3.10.2 the issue? 

I am trying to understand what changed from *yesterday to today*. Is it the python version, or some package I can pin? Basically, just a hint to what might have happened would be fantastic - these are tests so I have no problem pinning any package including "itsdangerous" (lol, pun intended, look at the above stacktrace, "itsdangerous" had a release 16 hours ago).

**Reply by potiuk** [reply]
Just look at the constraints. We have way more than 500 dependencies. We prepare constraints and manage them and if you have automated installation process that installs and reinstalls airflow you should use constraints to install the "golden" set of for sure working.

https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html

I have no idea what happened and when - when you have 500 dependencies every week you have more than 20 changes. and tracking it is impossible. I don't even know if this is itsdangerous that is causing the problems. 

Just follow the constraints way of installing airlfow (if you install it continuously). If you did install airlfow yesterday and did not follow this
https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html then you should follow it when you are installing Airflow. Just automate it to use our constraints.

**Reply by potiuk** [reply]
(and BTW in our constraints itsdangerous is 1.1.0 https://github.com/apache/airflow/blob/8a962740270a9ecf6b4f910173321b183e7adf4a/constraints-3.8.txt#L310 )

**Reply by potiuk** [reply]
Ah BTW. @tommyjcarpenter - since you last discussed it - we have a job in our CI/tool that now does SOME of what you wanted - in our "development" tools we have `breeze find-newer-dependencies` tool - https://github.com/apache/airflow/blob/main/BREEZE.rst#tracking-backtracking-issues-for-ci-builds that you can run in `main` and it can show you which of airflow dependencies changed recently (last few days). All you need is to checkout main and run it, and it will tell you list of dependencies tha have been released in the last few days.

This is as "close" as it can get when it comes to providing some clues of what changed (but constistently using constraints with released version is a way to have things stable so still it's a better advice).
